src/Swarm/Game/Step.hs

```diff
@@ -913,12 +919,52 @@ execConst c vs s k = do
         -- set the copied robot to be a "system" robot so it is
         -- capable of executing any commands; the As command
         -- already requires "God" capability.
-        v <-
-          evalState @Robot (r & systemRobot .~ True) . evalState @GameState g $
-            runCESK (Out prog s [FApp (VCApp Force []), FExec])
+
+        -- -----------------------------------------------------------------------------
+        -- TODO: move this code out
+        -- Slightly edited copy of Unit.hs version of runCESK:
+        let runCESK' :: CESK -> StateT Robot (StateT GameState IO) (Either Text Value)
+            runCESK' (Up exn _ []) = return (Left (formatExn exn))
+            runCESK' cesk = case finalValue cesk of
+              Just (v, _) -> return (Right v)
+              Nothing -> do
+                cesk1 <- stepCESK cesk
+                ticks += 1
+                runCESK' cesk1
+        -- -----------------------------------------------------------------------------
+        let evalAs :: IO (Either Text Value)
+            evalAs =
+              Out prog s [FApp (VCApp Force []), FExec]
+                & runCESK'
+                & flip evalStateT (r & systemRobot .~ True)
+                & flip evalStateT (g & gameMode .~ CreativeMode)
+
+        va <-
+          sendIO . Async.async $
+            either VString id -- TODO: that is not right...
+              <$> evalAs
 
         -- Return the value returned by the hypothetical command.
-        return $ Out v s k
+        i <- gensym <+= 1
+        return $ Out (VAsync $ GAsync i va) s (FApp (VCApp Poll []) : k)
+      _ -> badConst
+    Poll -> case vs of
+      [VAsync ga@(GAsync i va)] -> do
+        let fs :: Show a => a -> Text
+            fs = from . show
+        rn <- use robotName
+        time <- use ticks
+        -- DEBUG ------------------------------------------------------
+        mv <- sendIO $ Async.poll va
+        -- let mv = Nothing :: Maybe (Either SomeException Value)
+        let (msg, cesk) = case mv of
+              Nothing -> ("in progress", Waiting (time + 8) (Out (VAsync ga) s (FApp (VCApp Poll []) : k)))
```

Comment:
I wonder if it makes sense to use some kind of exponential backoff (up to a maximum), e.g. if it hasn't finished after 8 steps, then the next time we wait 16, then 32, ... up to a maximum of 64 or something like that?  For a long-running computation, checking every 8 time steps seems like a lot of polling, especially if we are writing to the robot log every time we check.  We could put the waiting time to use as an argument of `FPoll`, i.e. `FPoll n` means "check if the async computation is done, and if not, wait `n` time steps until checking again".

Reply:
I let `await` take `int` waiting period, but it would be fun to give it `int -> int` function instead :smile: